adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1226,11 +1185,75 @@ protected final void onCommandReceived(
+        if (tenantConfig.isAdapterEnabled(getTypeName())) {
+            LOG.debug("protocol adapter [{}] is enabled for tenant [{}]",
+                    getTypeName(), tenantConfig.getTenantId());
+            return Future.succeededFuture(tenantConfig);
+        } else {
+            LOG.debug("protocol adapter [{}] is disabled for tenant [{}]",
+                    getTypeName(), tenantConfig.getTenantId());
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN,
+                    "adapter disabled for tenant"));
+        }
+    }
+
+    private Future<DeviceUser> authenticateCredentials(final DeviceCredentials credentials) {
+        final Future<DeviceUser> result = Future.future();
+        usernamePasswordAuthProvider.authenticate(credentials, result.completer());
+        return result;
+    }
+
+    private Future<Device> isDeviceRegistered(final Device authenticatedDevice, final DeviceCredentials credentials,
+            final Span currentSpan) {
+        currentSpan.log(String.format("device authenticated"));
+        LOG.debug("successfully authenticated device [tenant-id: {}, auth-id: {}, device-id: {}]",
+                authenticatedDevice.getTenantId(), credentials.getAuthId(),
+                authenticatedDevice.getDeviceId());
+        return getRegistrationAssertion(authenticatedDevice.getTenantId(),
+                authenticatedDevice.getDeviceId(),
+                authenticatedDevice, currentSpan.context())
+                        .compose(ok -> {
+                            currentSpan.log(String.format("device registration verified"));
+                            LOG.debug("verified device registration [tenant-id: {}, device-id: {}]",
+                                    authenticatedDevice.getTenantId(), authenticatedDevice.getDeviceId());
+                            return accepted(authenticatedDevice);
+                        });
+    }
+
+    private Future<Device> createLink(final MqttEndpoint endpoint, final Device authenticatedDevice,
+            final Span currentSpan) {
+        final Future<MessageSender> telemetrySender = getTelemetrySender(authenticatedDevice.getTenantId());
+        final Future<MessageSender> eventSender = getEventSender(authenticatedDevice.getTenantId());
+
+        return CompositeFuture
+                .all(getRegistrationClient(authenticatedDevice.getTenantId()), telemetrySender, eventSender)
+                .map(ok -> {
+                    currentSpan.log(String.format("opened downstream links"));
+                    LOG.debug(
+                            "providently opened downstream links [credit telemetry: {}, credit event: {}] for tenant [{}]",
+                            telemetrySender.result().getCredit(), eventSender.result().getCredit(),
+                            authenticatedDevice.getTenantId());
+                    onAuthenticationSuccess(endpoint, authenticatedDevice);
```

Comment:
this should be invoked from within *handleEndpointConnectionWithAuthentication* as it is the last thing to do after the device has been authenticated etc.

Reply:
Handlers are being registered inside this method. Hence I have refactored this method name to `registerHandlers` and it will be invoked within `handleEndpointConnectionWithAuthentication`